python4j/python4j-numpy/src/main/java/NumpyArray.java

```diff
@@ -0,0 +1,146 @@
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ******************************************************************************/
+
+
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import org.apache.commons.lang3.ArrayUtils;
+import org.bytedeco.javacpp.Pointer;
+import org.nd4j.linalg.api.buffer.DataBuffer;
+import org.nd4j.linalg.api.buffer.DataType;
+import org.nd4j.linalg.api.concurrency.AffinityManager;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.api.shape.Shape;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.nativeblas.NativeOps;
+import org.nd4j.nativeblas.NativeOpsHolder;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.nd4j.linalg.api.buffer.DataType.FLOAT;
+
+
+/**
+ * Wrapper around INDArray for initializing from numpy array
+ *
+ * @author Fariz Rahman
+ */
+@Getter
+@NoArgsConstructor
+public class NumpyArray {
+
+    private static NativeOps nativeOps;
+    private static Map<String, INDArray> arrayCache;  // Avoids re-allocation of device buffer
+    private long address;
+    private long[] shape;
+    private long[] strides;
+    private DataType dtype;
+    private INDArray nd4jArray;
+
+    static {
+        //initialize
+        Nd4j.scalar(1.0);
+        nativeOps = NativeOpsHolder.getInstance().getDeviceNativeOps();
+        arrayCache = new HashMap<>();
+    }
+
+    @Builder
+    public NumpyArray(long address, long[] shape, long strides[], DataType dtype, boolean copy) {
+        this.address = address;
+        this.shape = shape;
+        this.strides = strides;
+        this.dtype = dtype;
+        setND4JArray();
+        if (copy) {
+            nd4jArray = nd4jArray.dup();
+            Nd4j.getAffinityManager().ensureLocation(nd4jArray, AffinityManager.Location.HOST);
+            this.address = nd4jArray.data().address();
+        }
+    }
+
+
+
+    public NumpyArray copy() {
+        return new NumpyArray(nd4jArray.dup());
+    }
+
+    public NumpyArray(long address, long[] shape, long strides[]) {
+        this(address, shape, strides, FLOAT, false);
+    }
+
+    public NumpyArray(long address, long[] shape, long strides[], DataType dtype) {
+        this(address, shape, strides, dtype, false);
+    }
+
+
+    private void setND4JArray() {
+
+        long size = 1;
+        for (long d : shape) {
+            size *= d;
+        }
+
+        String cacheKey = address + "_" + size + "_" + dtype + "_" + ArrayUtils.toString(strides);
+        nd4jArray = arrayCache.get(cacheKey);
```

Comment:
This looks a bit off... The cacheKey isn't guaranteed to be unique if the array is in a workspace (we could have the same address/size/dtype/strides in the next iteration).
One option is to ensure the array is always detached... in fact, doing the if statement content in a `try(MemoryWorkspace ws = Nd4j.getMemoryManager().scopeOutOfWorkspaces())` would be safest.

Or: alternatively use INDArray.getId() which is guaranteed to be unique even in the presence of workspaces.